src/backend/distributed/master/master_create_shards.c

```diff
@@ -217,8 +214,95 @@ master_create_worker_shards(PG_FUNCTION_ARGS)
 	}
 
 	RESUME_INTERRUPTS();
+}
 
-	PG_RETURN_VOID();
+
+/*
+ * CreateColocatedShards creates shards for the target relation colocated with
+ * the source relation.
+ */
+void
+CreateColocatedShards(Oid targetRelationId, Oid sourceRelationId)
+{
+	char *relationOwner = NULL;
+	char targetShardStorageType = 0;
+	List *existingShardList = NIL;
+	List *sourceShardList = NIL;
+	ListCell *sourceShardCell = NULL;
+
+	/* make sure that tables are hash partitioned */
+	CheckHashPartitionedTable(targetRelationId);
+	CheckHashPartitionedTable(sourceRelationId);
+
+	/*
+	 * In contrast to append/range partitioned tables it makes more sense to
+	 * require ownership privileges - shards for hash-partitioned tables are
+	 * only created once, not continually during ingest as for the other
+	 * partitioning types.
+	 */
+	EnsureTableOwner(targetRelationId);
+
+	/* we plan to add shards: get an exclusive metadata lock */
+	LockRelationDistributionMetadata(targetRelationId, ExclusiveLock);
+
+	relationOwner = TableOwner(targetRelationId);
+
+	/* validate that shards haven't already been created for this table */
+	existingShardList = LoadShardList(targetRelationId);
+	if (existingShardList != NIL)
+	{
+		char *targetRelationName = get_rel_name(targetRelationId);
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("table \"%s\" has already had shards created for it",
+							   targetRelationName)));
+	}
+
+	/* get shard storage type for the target relation */
+	targetShardStorageType = ShardStorageType(targetRelationId);
+
+	sourceShardList = LoadShardIntervalList(sourceRelationId);
+	foreach(sourceShardCell, sourceShardList)
+	{
+		ShardInterval *sourceShardInterval = (ShardInterval *) lfirst(sourceShardCell);
+		uint64 sourceShardId = sourceShardInterval->shardId;
+		List *sourceShardPlacementList = ShardPlacementList(sourceShardId);
+		Datum shardIdDatum = master_get_new_shardid(NULL);
+		int64 newShardId = DatumGetInt64(shardIdDatum);
```

Comment:
Might be worth putting the internals of master_get_new_shardid in a separate function that we can call here.